tls/s2n_config.c

```diff
@@ -840,6 +841,24 @@ int s2n_config_add_ticket_crypto_key(struct s2n_config *config,
     return 0;
 }
 
+int s2n_config_generate_local_ticket_crypto_key(struct s2n_config *config, uint64_t intro_time_in_seconds_from_epoch)
+{
+    s2n_stack_blob(key_name, S2N_TICKET_KEY_NAME_LEN, S2N_TICKET_KEY_NAME_LEN);
+    s2n_stack_blob(key, S2N_AES256_KEY_LEN, S2N_AES256_KEY_LEN);
+
+    POSIX_GUARD_RESULT(s2n_get_public_random_data(&key_name));
+    POSIX_GUARD_RESULT(s2n_get_private_random_data(&key));
+
+    POSIX_GUARD(s2n_config_add_ticket_crypto_key(config,
```

Comment:
Does this introduce any kind of mutability/threading issues?

If I remember correctly `s2_config` would be mostly immutable and could be shared between various threads if one uses other APIs which don't modify it. With this API it looks like it might not modify the config (it's just called `generate_local_ticket`, but in reality it does.

Or does `s2n_config_add_ticket_crypto_key` has internal synchronization on any kind of store - then this would also be fine.

Reply:
I think we would say this new API has the same thread safety guarantee as s2n_config_add_ticket_crypto_key, i.e. when you call generate but you should assume it will modify the s2n_config and your application needs to handle synchronization.

The recommendation would be to call generate during "building" the s2n_config and then the s2n_config would become immutable after that. There's also an interesting option for an application to "preload" multiple days worth of ticket keys, with the big downside of behavior changing unexpectedly when the preloaded tickets expire. Something like:

```
// number of days to preload session ticket keys
#define MAX_DAYS 14

for (int i = 0; i < MAX_DAYS; i++)
{
    # Generate a ticket key valid starting on day i
    s2n_config_generate_local_ticket_crypto_key(config, current_epoch_time_seconds + i * SECONDS_PER_DAY);
}

```